### PR TITLE
Add support for explicit broadcasting in TensorFlow

### LINKED ISSUE #14509: Function for explicit broadcasting
It would be convenient to have an explicit function for broadcasting in TensorFlow's Python API, like to [`numpy.broadcast_to`](https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.broadcast_to.html) or XLA's [Broadcast](https://www.tensorflow.org/performance/xla/operation_semantics#broadcast), and as [requested on StackOverFlow](https://stackoverflow.com/questions/34362193/how-to-explicitly-broadcast-a-tensor-to-match-anothers-shape-in-tensorflow). This would facilitate adding broadcasting-like behavior to the many TensorFlow operations that don't support it out of the box.

I understand that in general TensorFlow does not implement NumPy's the strided N-dimensional array data model, so unlike the case for NumPy, broadcasting (e.g., with Eigen tensors) can require a copy. This is a good reason to not necessarily build such a version of broadcasting into ops. However, explicit broadcasting rather than using tile/expand_dims can still be very convenient.

### PR DESCRIPTION
**Pull request by yongtang**
This fix tries to adds support for explicit broadcasting in TensorFlow, as was suggested in #14509. This fix adds the op of tf.broadcast_to, which is equivalent to the numpy.broadcast_to in numpy.

This fix fixes #14509.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

### CODE REVIEW

**Comment by caisq** [comment]
@rmlarsen Do you want to review this CL? @martinwicke I assume this requires an API review.

File: tensorflow/core/ops/array_ops.cc [hunk 1]
```diff
@@ -410,6 +410,20 @@ axis: Dimension along which to unpack.  Negative values wrap around, so the
 output: The list of tensors unpacked from `value`.
 )doc");
 
+REGISTER_OP("BroadcastTo")
+    .Input("input: T")
+    .Input("shape: int32")
```
Comment (alextp) [inline tensorflow/core/ops/array_ops.cc]: Might be worth it to support int64 shapes as well

File: tensorflow/core/ops/array_ops.cc [hunk 2]
```diff
@@ -410,6 +410,20 @@ axis: Dimension along which to unpack.  Negative values wrap around, so the
...
+    .Output("output: T")
+    .Attr("T: type")
+    .SetShapeFn([](InferenceContext* c) { return Status::OK(); })
```
Comment (alextp) [inline tensorflow/core/ops/array_ops.cc]: You can do much more than a pass-through shape inference function here.

Specifically, if the shape tensor has a known constant value (and it often will) you should be able to set the output shape to be exactly that shape. See the shape_fn for reshape; you might be able to reuse most of its code here.

File: tensorflow/core/kernels/broadcast_to_op.cc [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+
+    switch (output_shape.dims()) {
+      case 1: {
+        auto reshape = AsEigenDSizesWithPrefix<1>(input_shape);
+        auto broadcast = output_shape.AsEigenDSizes<1>();
+
+        BROADCAST_SHAPE(broadcast, reshape, 1, input_shape, output_shape);
+
+        auto output = output_tensor->tensor<T, 1>();
+        switch (input_shape.dims()) {
```
Comment (alextp) [inline tensorflow/core/kernels/broadcast_to_op.cc]: Not allowing broadcasting from a scalar?

File: tensorflow/python/kernel_tests/array_ops_test.py [hunk 4]
```diff
@@ -1123,6 +1123,40 @@ def testResourceRejection(self):
                                                "cannot be a resource variable"):
         guarantee_a.eval()
 
+class BroadcastToTest(test_util.TensorFlowTestCase):
```
Comment (alextp) [inline tensorflow/python/kernel_tests/array_ops_test.py]: I think it'd be nice to have these tests in their own file, as this file is already quite long.

**Review by alextp — CHANGES_REQUESTED** [review]
Overall looks good, some minor comments

File: tensorflow/core/kernels/broadcast_to_op.cc (hunk 3 above)
Comment (yongtang) [reply tensorflow/core/kernels/broadcast_to_op.cc]: Thanks @alextp. The PR has been updated to processing scalar with `setConstant` to broadcast.

File: tensorflow/core/ops/array_ops.cc (hunk 1 above)
Comment (yongtang) [reply tensorflow/core/ops/array_ops.cc]: Thanks @alextp. Done.

File: tensorflow/core/ops/array_ops.cc (hunk 2 above)
Comment (yongtang) [reply tensorflow/core/ops/array_ops.cc]: Done. Thanks.

File: tensorflow/python/kernel_tests/array_ops_test.py (hunk 4 above)
Comment (yongtang) [reply tensorflow/python/kernel_tests/array_ops_test.py]: Done. Thanks.

**Comment by yongtang** [comment]
@alextp Thanks for the review. The PR has been updated. Please take a look and let me know if there are any issues.

File: tensorflow/core/kernels/broadcast_to_op.cc [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+    for (int d = NDIMS - shape.dims(); d < NDIMS; d++) {
+      dsizes[d] = shape.dim_size(d - (NDIMS - shape.dims()));
+    }
+    return dsizes;
+  }
+};
+
+// As MakeShape is able to handle both DT_INT32 and DT_INT64,
+// no need to have TypeConstraint for `Tidx`
+#define REGISTER_KERNEL(type)                                           \
```
Comment (alextp) [inline tensorflow/core/kernels/broadcast_to_op.cc]: Register a GPU kernel as well? That might require adding a .device() before the eigen call to .broadcast but otherwise should be easy.

File: tensorflow/core/ops/array_ops.cc [hunk 6]
```diff
@@ -410,6 +410,65 @@ axis: Dimension along which to unpack.  Negative values wrap around, so the
...
+
+      c->set_output(0, out);
+      return Status::OK();
+    })
+    .Doc(R"doc(
+Repeat elements of an array
+
+input: A Tensor to broadcast.
+shape: An 1-D `int` Tensor. The shape of the desired output.
+output: A Tensor.
```
Comment (alextp) [inline tensorflow/core/ops/array_ops.cc]: It might be useful to explain here how broadcasting works (align dimensions, 1s are expanded to be the size, etc)

File: tensorflow/python/kernel_tests/broadcast_to_ops_test.py [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+from tensorflow.python.framework import test_util
+from tensorflow.python.ops import array_ops
+from tensorflow.python.platform import test as test_lib
+
+
+class BroadcastToTest(test_util.TensorFlowTestCase):
+
+  def testBroadcastToBasic(self):
+    for dtype in [np.uint8, np.uint16, np.int8, np.int16, np.int32, np.int64]:
+      with self.test_session():
```
Comment (alextp) [inline tensorflow/python/kernel_tests/broadcast_to_ops_test.py]: You want self.test_session(use_gpu=True) to make sure the GPU gets used when it's available (it's a no-op when the GPU is not available)

**Review by alextp — CHANGES_REQUESTED** [review]
Thanks for the changes!

File: tensorflow/python/kernel_tests/broadcast_to_ops_test.py [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+        # check shape inference when shape input is constant
+        self.assertAllEqual(shape, v_np.shape)
+
+  def testBroadcastToScalar(self):
+    with self.test_session():
+      x = np.array(1, dtype=np.int32)
+      v_tf = array_ops.broadcast_to(constant_op.constant(x), [3, 3])
+      v_np = np.broadcast_to(x, [3, 3])
+      self.assertAllEqual(v_tf.eval(), v_np)
+
```
Comment (shoyer) [inline tensorflow/python/kernel_tests/broadcast_to_ops_test.py]: Please take a look at NumPy's tests for `broadcast_to` as an example of what good coverage for shapes looks like:
https://github.com/numpy/numpy/blob/4f4b0a43c2c5c1441a507a214045a50a1bfc03ed/numpy/lib/tests/test_stride_tricks.py#L222-L265

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by martinwicke** [comment]
@av8ramit Is this nag (wording) appropriate since this has "awaiting response"?

**Comment by av8ramit** [comment]
@martinwicke yeah might need to toggle the state machine a bit.

**Comment by martinwicke** [comment]
@yongtang can you look at the latest comments?

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.

**Comment by yongtang** [comment]
The doctoring has been updated. Will find a GPU machine to test and enable GPU kernel.

File: tensorflow/core/kernels/BUILD [hunk 9]
```diff
@@ -606,6 +606,7 @@ cc_library(
         ":batch_space_ops",
         ":bcast_ops",
         ":bitcast_op",
+        ":broadcast_to_op",
```
Comment (alextp) [inline tensorflow/core/kernels/BUILD]: Why is this here?

File: tensorflow/python/ops/array_ops.py [hunk 10]
```diff
@@ -28,6 +28,7 @@
 @@saturate_cast
 @@broadcast_dynamic_shape
 @@broadcast_static_shape
+@@broadcast_to
```
Comment (alextp) [inline tensorflow/python/ops/array_ops.py]: Let's add this to contrib for one release before we commit to supporting this API forever. To do so, remove this @@ annotation here, revert the change to tensorflow.pbtxt, and add visibility: HIDDEN to the apidef.pbtxt proto. Then, import this symbol (broadcast_to) from https://github.com/tensorflow/tensorflow/blob/master/tensorflow/contrib/framework/__init__.py so it gets exposed as tf.contrib.framework.broadcast_to

**Review by alextp — CHANGES_REQUESTED** [review]

File: tensorflow/python/ops/array_ops.py (hunk 10 above)
Comment (martinwicke) [reply tensorflow/python/ops/array_ops.py]: Can't you now simply do tf_export('tf.contrib.framework.broadcast_to')?

@annarev?

File: tensorflow/core/kernels/BUILD (hunk 9 above)
Comment (yongtang) [reply tensorflow/core/kernels/BUILD]: @alextp The dependency is added so that it is part of the `gen_array_ops.py`.

File: tensorflow/python/ops/array_ops.py (hunk 10 above)
Comment (yongtang) [reply tensorflow/python/ops/array_ops.py]: @alextp @martinwicke I add `tf_export('contrib.framework.broadcast_to')(broadcast_to)` and that seems to do the work. Although I also need to update `_allowed_symbols` in `tensorflow/contrib/framework/__init__.py` so that the symbol could be exposed.

**Review by alextp — APPROVED** [review]

File: tensorflow/python/ops/array_ops.py (hunk 10 above)
Comment (yongtang) [reply tensorflow/python/ops/array_ops.py]: Thanks @annarev. I updated the PR so that it is exposed from `tensorflow/contrib/framework/__init__.py`.

**Comment by hadim** [comment]
Is that normal I can't see `tf.broadcast_to` in the last 1.8.0 version?

**Comment by Hoeze** [comment]
I just compiled the "r1.9" branch, but there is still no "tf.broadcast_to":
```
>>> tf.__version__
'1.9.0-rc0'
>>> tf.broadcast_to
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'tensorflow' has no attribute 'broadcast_to'
```
How can I get a Tensorflow-Version supporting this?

**Comment by martinwicke** [comment]
This PR added the op, but hid it by adding this file: tensorflow/core/api_def/python_api/api_def_BroadcastTo.pbtxt

Not sure whether that was intentional, but removing that file should make this op visible.

**Comment by Hoeze** [comment]
Removing means deleting it?

**Comment by martinwicke** [comment]
I think that would be sufficient.​

**Comment by Hoeze** [comment]
Thank you very much, deleting that file and recompiling everything did the trick!

**Comment by yongtang** [comment]
@Hoeze The op is exposed as `tf.contrib.framework.broadcast_to` for now (See https://github.com/tensorflow/tensorflow/pull/15243#discussion_r178889285):
```
# python
Python 2.7.12 (default, Dec  4 2017, 14:50:18) 
[GCC 5.4.0 20160609] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import tensorflow as tf
>>> tf.__version__
'1.9.0-dev20180604'
>>> tf.contrib.framework.broadcast_to
<function broadcast_to at 0x7f1f8588c230>
>>> 
```

**Comment by martinwicke** [comment]
​Could you send a PR? I don't see a reason to hide this op.

**Comment by yongtang** [comment]
@martinwicke @Hoeze The PR #19753 has been created to expose `tf.broadcast_to`.

**Comment by Hoeze** [comment]
@yongtang This patch should be merged to `r1.9` branch as well, shouldn't it?
